Add regex support for pull request links

ephemeral8997 opened the pull request:
This pr will add message edition support to the pull request index detecting, and it uses regex.

ephemeral8997 committed:
Use regex instead of looping

Regex is more optimised than looping

ephemeral8997 committed:
Support message edition 

If you send an incorrect pull request index by mistake, you can edit the message and bot content will be edited too. This works for the last 100 messages in the last hour since the message has been sent.

ephemeral8997 committed:
Fix bugs and typos

The message edition support has been added now with a better regex.

PythonIntermediateCoder commented:
The pr now gets the last 10 messages instead, because we do not want to spam the API with useless HTTP calls. Also, there're regexes for pull request link and its index.

Dorukyum commented:
I think we can put the edit handling on hold for now but the regex search is perfect.

Dorukyum commented on cogs/general.py:
```diff
@@ -2,12 +2,16 @@
 from contextlib import suppress
 from urllib import parse
+import re
+import datetime
```
Why is datetime imported?

ephemeral8997 replied:
I forgot to remove it when I wanted to use datetimes to determine the message to edit, I'll remove it

ephemeral8997 committed:
Update general.py

PythonIntermediateCoder commented:
The message edition needs a bit improvement, It makes a HTTP request per message edition so It takes time and abuses the API when editing for many times

Dorukyum commented:
That's why I think we should put it on hold for now.

Dorukyum commented on cogs/general.py:
```diff
@@ -2,12 +2,15 @@
 from utils import Cog
 
+PULL_HASH_REGEX = re.compile(r'##(?P<index>[0-9]+)')
+GITHUB_URL_REGEX = re.compile(r'http(s)?://(www\.)?github\.com/Pycord-Development/pycord/pull/(?P<index>[0-9]+)')
```
This is used by gitlink now so we should remove it from here
```suggestion
```

Dorukyum commented on cogs/general.py:
```diff
@@ -74,14 +77,46 @@ async def on_message(self, message: discord.Message):
         if links:
             await message.reply("\n".join(links))
 
+    @Cog.listener()
+    async def on_message_edit(self, before: discord.Message, after: discord.Message):
+        github_url_format = "https://github.com/Pycord-Development/pycord/pull/{index}" # a template for shorter code
+        if before.author.bot:
+            return
+        before_pull_indexes = PULL_HASH_REGEX.findall(before.content)
+        after_pull_indexes = PULL_HASH_REGEX.findall(after.content)
+        if not before_pull_indexes:
+            # it is a normal message
+            return
+        if after_pull_indexes == before_pull_indexes:
+            # pull indexes have not been changed yet
+            return
+        created_at = before.created_at
+        async for message in before.channel.history(limit=5, after=created_at):
+            if message.author.id != self.bot.user.id:
+                continue
+            content = message.content
+            all_links = list(link[2] for link in GITHUB_URL_REGEX.findall(content))
+            # we will modify the message content
+            # so it matches the new  pull indexes
+            splitted = content.split()
+            new_list = []
+            for word in splitted:
+                for link in all_links:
+                    try:
+                        link = github_url_format.format(index=after_pull_indexes[before_pull_indexes.index(link)])
+                    except IndexError:
+                        # new pull index added
+                        new_list.append(link)
+                    else:
+                        if link not in new_list:
+                            new_list.append(link)
+            await message.edit(content='\n'.join(new_list))
+            return
+
```
I really dont think we should do this, at least not yet
```suggestion
```

ephemeral8997 replied:
I'll make it as a null docstring to leave it.

ephemeral8997 replied:
or maybe I'll just remove it so we can do it better in another time.

ephemeral8997 committed:
Remove message edition support for now

Current on_message_edit is not optimised as It takes more time to do its tasks since it makes a HTTP call per edition.

Co-authored-by: Dorukyum <53639936+Dorukyum@users.noreply.github.com>

ephemeral8997 committed:
Remove unnecessary regex

Co-authored-by: Dorukyum <53639936+Dorukyum@users.noreply.github.com>

ephemeral8997 committed:
Update general.py

ephemeral8997 committed:
Merge branch 'main' into patch-3

Dorukyum committed:
Update general.py
